http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpFullRequest.java

```diff
@@ -66,9 +71,29 @@
         @Override
         default Builder uri(URI uri) {
             return this.protocol(uri.getScheme())
-                       .host(uri.getHost())
-                       .port(uri.getPort())
-                       .encodedPath(SdkHttpUtils.appendUri(uri.getRawPath(), encodedPath()));
+                    .host(uri.getHost())
+                    .port(uri.getPort())
+                    .encodedPath(SdkHttpUtils.appendUri(uri.getRawPath(), encodedPath()));
+        }
+
+        /**
+         * Convenience method to set the {@link #protocol()}, {@link #host()}, {@link #port()},
+         * {@link #encodedPath()} and keep query parameters if present from a {@link URI} object.
+         *
+         * @param uri URI containing protocol, host, port, path and optional query parameters.
+         * @return This builder for method chaining.
+         */
+        default Builder uriWithQueryParams(URI uri) {
+            Builder builder = this.uri(uri);
+            if (uri.getQuery() != null) {
+                Pattern.compile("\\s*&|%26\\s*")
```

Comment:
- I don't think we should including `%26` here as the literal `&` should be the only delimiter we look for.
 - Compiling a pattern each time is inefficient. Once this logic is moved into `SdkHttpUtils`, let's extract the pattern out into a static constant

Reply:
In appendUriQueryParameter impl I now decode the incoming paramValue as was suggested before, which is why I reasoned to also split on the decoded delimiter as otherwise the value would not be split properly before appending?

Reply:
Unfortunately I don't think that's the right strategy. We should do the decoding after we have the individual components of the query parsed out. Otherwise, if the the input `URI` has and encoded `&` or `=`, then we would be parsing it incorrectly.  It's perfectly valid for the parameter name or value to have `&` or `=` it as long as it's encoded.

For example, the pair [contains=equals, foo] in a URI would be `https://foo.bar?contains%3Dequals=foo`, but the code as is will split this as `[contains, equals=foo]`

Reply:
I think the general strategy should be

 - get components of query string
 - foreach component
   - decode name
   - decode value if present
 - put raw query param

Reply:
I haven't thought about it in that way, makes sense to split like that and then decode both parts before adding.

On reusing the put raw query param bullet, that would mean we cannot differ on where te params come from as I did now in:

    private Map<String, List<String>> determineQueryParameters(Builder builder) {
        if (builder.uriQueryParameters.isEmpty()) {
            return builder.queryParametersAreFromToBuilder
                   ? builder.queryParameters
                   : deepUnmodifiableMap(builder.queryParameters, () -> new LinkedHashMap<>());
        }
        return builder.uriQueryParameters;
    }

Reply:
Hmm I don't think it should be necessary to distinguish the source. `uriWithQueryParams` acts basically like a static constructor for `SdkHttpRequest`

Reply:
Attempted to reflect this train of thought